pkg/planner/core/plan_cache_instance.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"go.uber.org/atomic"
+)
+
+// InstancePlanCache represents the instance/node level plan cache.
+type InstancePlanCache interface {
+	// Get gets the cached value from the cache according to key and opts.
+	Get(sctx sessionctx.Context, key string, opts any) (value any, ok bool)
+	// Put puts the key and value into the cache.
+	Put(sctx sessionctx.Context, key string, value, opts any) (succ bool)
```

Comment:
Maybe we should add a comment to explain why we need to use them here. But I am fine if you prefer not to add it.

Reply:
Added, Thanks!